kore/src/Kore/Simplify/Condition.hs

```diff
@@ -78,9 +81,26 @@ simplify ::
     SideCondition RewritingVariableName ->
     Conditional RewritingVariableName any ->
     LogicT simplifier (Conditional RewritingVariableName any)
-simplify SubstitutionSimplifier{simplifySubstitution} sideCondition =
-    normalize >=> worker
+simplify SubstitutionSimplifier{simplifySubstitution} sideCondition original =
+    normalize original >>= loop 0
   where
+    limit :: Int
+    limit = 4
+
+    loop ::
+        Int ->
+        Conditional RewritingVariableName any ->
+        LogicT simplifier (Conditional RewritingVariableName any)
+    loop count input
+        | count >= limit = do
+            warnUnsimplifiedCondition limit original input
+            pure input
+        | otherwise = do
+            output <- worker input
+            if fullySimplified output
+                then return (extract output)
+                else loop (count + 1) (extract output)
```

Comment:
Instead of explicit recursion can we use `foldM`?:
```suggestion
simplify SubstitutionSimplifier{simplifySubstitution} sideCondition original = do
    normalizedOriginal <- normalize original
    (isFullySimplified, result) <- foldM simplifyingCondition (False, normalizedOriginal) [0..limit]
    unless isFullySimplified $ warnUnsimplifiedCondition limit original result
    return result
  where
    limit :: Int
    limit = 4
    
    simplifyingCondition ::
        (Bool, Conditional RewritingVariableName any) ->
        Int ->
        LogicT simplifier (Bool, Conditional RewritingVariableName any)
    simplifyingCondition result@(True, _) _ = return result
    simplifyingCondition (_, input) _ = do
        output <- worker input
        return (fullySimplified output, extract output)
```

Reply:
This also looks fine, although I'd probably have spent more time trying to understand it if I didn't know it's meant to replace the initial recursion.

Reply:
This was my initial thought as well, the code isn't as straightforward with `foldM`. It also seems harder to debug, and we'll inevitably have lots of occasions to debug this I'm afraid, because of the issues with the simplifier. @dopamane, what do you think?

Reply:
TODO: add comment linking to old version